planner/cascades/transformation_rules.go

```diff
@@ -2502,3 +2505,127 @@ func (r *PullSelectionUpApply) OnTransform(old *memo.ExprIter) (newExprs []*memo
 	newApplyGroupExpr.SetChildren(outerChildGroup, old.Children[1].GetExpr().Children[0])
 	return []*memo.GroupExpr{newApplyGroupExpr}, false, false, nil
 }
+
+// MergeAdjacentWindow merge adjacent Window.
+type MergeAdjacentWindow struct {
+	baseRule
+}
+
+// NewRuleMergeAdjacentWindow creates a new Transformation MergeAdjacentWindow.
+// The pattern of this rule is `Window -> Window`.
+func NewRuleMergeAdjacentWindow() Transformation {
+	rule := &MergeAdjacentWindow{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandWindow,
+		memo.EngineAll,
+		memo.NewPattern(memo.OperandWindow, memo.EngineAll),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *MergeAdjacentWindow) Match(expr *memo.ExprIter) bool {
+	curWinPlan := expr.GetExpr().ExprNode.(*plannercore.LogicalWindow)
+	nextWinPlan := expr.Children[0].GetExpr().ExprNode.(*plannercore.LogicalWindow)
+	ctx := expr.GetExpr().ExprNode.SCtx()
+
+	// Whether Partition parts are the same.
+	if len(curWinPlan.PartitionBy) != len(nextWinPlan.PartitionBy) ||
+		len(curWinPlan.OrderBy) != len(nextWinPlan.OrderBy) ||
+		(curWinPlan.Frame == nil && nextWinPlan.Frame != nil) ||
+		(curWinPlan.Frame != nil && nextWinPlan.Frame == nil) {
+		return false
+	}
+	for i, item := range curWinPlan.PartitionBy {
+		if !item.Col.Equal(ctx, nextWinPlan.PartitionBy[i].Col) ||
+			item.Desc != nextWinPlan.PartitionBy[i].Desc {
+			return false
+		}
+	}
+
+	// Whether OrderBy parts are the same.
+	for i, item := range curWinPlan.OrderBy {
+		if !item.Col.Equal(ctx, nextWinPlan.OrderBy[i].Col) ||
+			item.Desc != nextWinPlan.OrderBy[i].Desc {
+			return false
+		}
+	}
+
+	// Whether Frame parts are the same.
+	if curWinPlan.Frame != nil {
+		if curWinPlan.Frame.Type != nextWinPlan.Frame.Type ||
+			curWinPlan.Frame.Start.Type != nextWinPlan.Frame.Start.Type ||
+			curWinPlan.Frame.Start.UnBounded != nextWinPlan.Frame.Start.UnBounded ||
+			curWinPlan.Frame.Start.Num != nextWinPlan.Frame.Start.Num {
+			return false
+		}
+		for i, expr := range curWinPlan.Frame.Start.CalcFuncs {
+			if !expr.Equal(ctx, nextWinPlan.Frame.Start.CalcFuncs[i]) {
+				return false
+			}
+		}
+		for i, expr := range curWinPlan.Frame.End.CalcFuncs {
+			if !expr.Equal(ctx, nextWinPlan.Frame.End.CalcFuncs[i]) {
+				return false
+			}
+		}
+	}
+
+	// Whether the first window uses the unsettled columns in the next window.
+
+	// `select a, b, sum(bb) over (partition by a) as 'sum_bb' from (select a, b, max(b) over (partition by a) as 'bb' from t) as tt`
+	// The adjacent windows in the above sql statement cannot be merged.
+	// The reason is that the first one uses an unsettled column `bb` from the second one.
+	nextWindowNewCols := make(map[*expression.Column]struct{})
+	for _, nc := range nextWinPlan.Schema().Columns {
```

Comment:
You should use the schema of the `Group` instead of the `LogicalPlan`.

Reply:
Is the schema in `Group` the same as in `LogicalPlan` 😄?

Reply:
No, they are not equivalent. 

In cascades planner, the schema of `LogicalPlan` should never be used. We don't set the schema of new created `LogicalPlan`s. Every time we want to set or get the schema, we use the schema of `Group` instead.